apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/data_updater/impl.ex

```diff
@@ -87,15 +97,78 @@ defmodule Astarte.DataUpdaterPlant.DataUpdater.Impl do
     Logger.metadata(realm: realm, device_id: encoded_device_id)
     Logger.info("Created device process.", tag: "device_process_created")
 
-    {:ok, db_client} = Database.connect(realm: new_state.realm)
+    {:ok, db_client} = Database.connect(realm: state.realm)
 
     stats_and_introspection =
       Queries.retrieve_device_stats_and_introspection!(db_client, device_id)
 
     {:ok, ttl} = Queries.fetch_datastream_maximum_storage_retention(db_client)
 
-    Map.merge(new_state, stats_and_introspection)
-    |> Map.put(:datastream_maximum_storage_retention, ttl)
+    new_state =
+      Map.merge(state, stats_and_introspection)
+      |> Map.put(:datastream_maximum_storage_retention, ttl)
+
+    {:ok, new_state}
+  end
+
+  @impl true
+  def handle_message(payload, headers, message_id, timestamp, state) do
+    %{@msg_type_header => message_type} = headers
+
+    case message_type do
+      "connection" ->
+        %{@ip_header => ip_address} = headers
+
+        handle_connection(state, ip_address, timestamp)
+
+      "disconnection" ->
+        handle_disconnection(state, timestamp)
+
+      "heartbeat" ->
+        handle_heartbeat(state, timestamp)
+
+      "internal" ->
+        %{@internal_path_header => internal_path} = headers
+        handle_internal(state, internal_path, payload, timestamp)
+
+      "introspection" ->
+        handle_introspection(state, payload, timestamp)
+
+      "data" ->
+        %{@interface_header => interface, @path_header => path} = headers
+        handle_data(state, interface, path, payload, timestamp)
+
+      "control" ->
+        %{@control_path_header => control_path} = headers
+        handle_control(state, control_path, payload, timestamp)
+
+      _ ->
+        # ack to discard unhandled messages
+        {:ack, :ok, state}
+    end
+  end
+
+  @impl Handler
+  def handle_signal(signal, state) do
+    case signal do
+      {:handle_install_volatile_trigger, object_id, object_type, parent_id, simple_trigger_id,
```

Comment:
this is the only `handle_` function where we pattern match on the argument inside the function, I'd keep this at function signature level for consistency but don't feel strongly about it